crates/bevy_reflect/src/serde/de.rs

```diff
@@ -12,375 +23,565 @@ pub trait DeserializeValue {
+    type Value = DynamicStruct;
 
-    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+        formatter.write_str("reflected struct value")
+    }
+
+    fn visit_map<V>(self, mut map: V) -> Result<Self::Value, V::Error>
     where
-        D: serde::Deserializer<'de>,
+        V: MapAccess<'de>,
     {
-        deserializer.deserialize_seq(ListVisitor {
-            registry: self.registry,
-        })
+        visit_struct(&mut map, self.struct_info, self.registry)
     }
 }
 
-struct ListVisitor<'a> {
+struct TupleStructVisitor<'a> {
+    tuple_struct_info: &'static TupleStructInfo,
     registry: &'a TypeRegistry,
 }
 
-impl<'a, 'de> Visitor<'de> for ListVisitor<'a> {
-    type Value = DynamicList;
+impl<'a, 'de> Visitor<'de> for TupleStructVisitor<'a> {
+    type Value = DynamicTupleStruct;
 
-    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
-        formatter.write_str("list value")
+    fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
+        formatter.write_str("reflected tuple struct value")
     }
 
     fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
     where
         V: SeqAccess<'de>,
     {
-        let mut list = DynamicList::default();
-        while let Some(value) = seq.next_element_seed(ReflectDeserializer {
+        let mut index = 0usize;
+        let mut tuple_struct = DynamicTupleStruct::default();
+
+        let get_field_registration = |index: usize| -> Result<&'a TypeRegistration, V::Error> {
+            let field = self.tuple_struct_info.field_at(index).ok_or_else(|| {
+                de::Error::custom(format_args!(
+                    "no field at index {} on tuple {}",
+                    index,
+                    self.tuple_struct_info.type_name(),
+                ))
+            })?;
+            get_registration(field.type_id(), field.type_name(), self.registry)
+        };
+
+        while let Some(value) = seq.next_element_seed(TypedReflectDeserializer {
+            registration: get_field_registration(index)?,
             registry: self.registry,
         })? {
-            list.push_box(value);
+            tuple_struct.insert_boxed(value);
+            index += 1;
+            if index >= self.tuple_struct_info.field_len() {
+                break;
+            }
         }
-        Ok(list)
+
+        let ignored_len = self
+            .registry
+            .get(self.tuple_struct_info.type_id())
+            .and_then(|registration| registration.data::<SerializationData>())
+            .map(|data| data.len())
+            .unwrap_or(0);
+        if tuple_struct.field_len() != self.tuple_struct_info.field_len() - ignored_len {
+            return Err(Error::invalid_length(
+                tuple_struct.field_len(),
+                &self.tuple_struct_info.field_len().to_string().as_str(),
+            ));
+        }
```

Comment:
Thoughts on adding this registry lookup so we can provide a proper error message? Invalid lengths are important to catch as letting it pass silently might cause issues further down the line for users.

Reply:
I'm on board, but we could optimize this by passing in the TypeRegistration reference to TupleStructVisitor to avoid looking it up again.

Reply:
Done!